CMakeLists.txt

```diff
@@ -0,0 +1,32 @@
+cmake_minimum_required(VERSION 3.10)
+
+project(UriParser VERSION 0.9.1 LANGUAGES C)
+
+file(GLOB INTERFACE_FILES
+        RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}" CONFIGURE_DEPENDS
+        "include/uriparser/*.h")
+
+file(GLOB SOURCE_FILES
+        RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}" CONFIGURE_DEPENDS
+        "src/*.[c|h]")
+
+add_library(uriparser ${INTERFACE_FILES} ${SOURCE_FILES})
+set_property(TARGET uriparser PROPERTY PUBLIC_HEADER "${INTERFACE_FILES}")
+target_include_directories(uriparser PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_include_directories(uriparser PUBLIC $<INSTALL_INTERFACE:include>)
+
+install(TARGETS uriparser
+        EXPORT uriparser
+        LIBRARY
+            DESTINATION lib
+        ARCHIVE
+           DESTINATION lib
+        PUBLIC_HEADER
+            DESTINATION include/uriparser
+        )
+
+install(EXPORT uriparser
+        DESTINATION lib/cmake/uriparser
+        NAMESPACE UriParser::
+        )
```

Comment:
Could you elaborate what this last section is for?

Reply:
When you call `find_package(XXX)` in a cmakelists file, cmake will look in a number of places, first it will look for a module called `FindXXX.cmake` in the `CMAKE_MODULE_PATH`. If that's not found, it will search the `CMAKE_PREFIX_PATH` to search for the files that this line creates. In the following example I ran cmake with `-DCMAKE_INSTALL_PREFIX=${HOME}/install/native-cxx17/uriparser/Release`
The output goes here:
```
$ ls -la lib/cmake/uriparser/
total 16
drwxr-xr-x  4 richardhodges  richardhodges   128 Jan  7 13:06 .
drwxr-xr-x  3 richardhodges  richardhodges    96 Jan  7 12:20 ..
-rw-r--r--  1 richardhodges  richardhodges   865 Jan  7 13:06 uriparser-release.cmake
-rw-r--r--  1 richardhodges  richardhodges  3271 Jan  7 11:47 uriparser.cmake
```

and

```
$ ls -la lib/
total 264
drwxr-xr-x  4 richardhodges  richardhodges     128 Jan  7 13:06 .
drwxr-xr-x  4 richardhodges  richardhodges     128 Jan  7 13:06 ..
drwxr-xr-x  3 richardhodges  richardhodges      96 Jan  7 12:20 cmake
-rw-r--r--  1 richardhodges  richardhodges  132776 Jan  7 13:06 liburiparser.a
```

This means that anyone building this package can now simply install into their own directory, set URIPARSER_DIR to `{prefix}/lib/cmake/uriparser` and then
```
find_package(uriparser CONFIG REQUIRED)
```
The URIPARSER_DIR step is unnecessary if it's been installed into a standard place (e.g. /ur/local/)

This will mean that the script importing uriparser can simply:
```
find_package(uriparser CONFIG REQUIRED)
target_link_libraries(my-target PUBLIC UriParser::uriparser) # namespace::target
```

And all include/library dependencies will be both set up for this target and propagated to any other target that uses `my-target`.

Reply:
In the project for which I have used uriparser, I actually use `ExternalProject_Add` for now. One of the problems with that is that you don't get the luxury of being able to use the packaging system (because the targets resulting from ExternalProject_Add are built at build time and cmake's find_package needs to resolve packages at configuration time).

In this case I simply build a mock target - which is essentially what the find_package code is doing:

```
ExternalProject_Add(EP_URIPARSER
        PREFIX "${CMAKE_CURRENT_BINARY_DIR}"
        GIT_REPOSITORY "https://github.com/QuoineFinancial/uriparser.git"
        GIT_TAG "master"
        GIT_PROGRESS TRUE
        CMAKE_ARGS
            -DCMAKE_TOOLCHAIN_FILE:FILEPATH=${CMAKE_TOOLCHAIN_FILE}
            -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
            -DCMAKE_BUILD_TYPE:STRING=Release
        )
ExternalProject_Get_Property(EP_URIPARSER INSTALL_DIR)
set(uriparser_install_dir "${INSTALL_DIR}")
add_library(uriparser-dependency INTERFACE)
add_dependencies(uriparser-dependency EP_URIPARSER)
target_include_directories(uriparser-dependency SYSTEM INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>")
target_link_libraries(uriparser-dependency INTERFACE "$<BUILD_INTERFACE:uriparser>")
target_link_directories(uriparser-dependency INTERFACE "$<BUILD_INTERFACE:${uriparser_install_dir}/lib>")
add_library(UriParser::uriparser ALIAS uriparser-dependency)

```

Reply:
I don't get all of it yet but many thanks for the explanation! I have added some export bytes to #47 now.